### DISCUSSION TITLE
What's the correct way to add resources to an existing one?

### DISCUSSION THREAD

**Discussion by johnkwaters** [FirelyTeam/firely-net-sdk#2085] [Q&A]
If I fetch a patient, and then want to send back some changes - for instance, I change the Patient's address and add an Encounter with some vitals... what is the correct way to do this?
Do I firsts fetch the patient record to get the latest version, then modify that patient.Address, then add that patient and a new encounter with a new observation as three resources to a bundle?
Do I use transaction builder?
I am not finding any  documentation that walks through how to deal with these kinds of partial update and additions?

**Comment by mbaltus** [comment]
To update the patient, the best practice is indeed to request the current data first, then make changes and send an update request to the server. Or you could do a PATCH instead, if you do not want to fetch the patient first. Since you mention 'partial update', it would be good to take a look at PATCH (http://hl7.org/fhir/http.html#patch) and see if your server supports it.

For the other additions, make sure they have the references set correctly, so they will get connected to your patient and each other. Then you can either do separate POST calls for each, or create a transaction Bundle with the transaction builder. Which one is the best approach will depend on a number of things.
Transactions are useful when
- the server supports transaction handling,
- the resources have references to each other, and you want the server to handle updating those after the resources have been created instead of doing that one by one
- you want to use conditional references in the resources, which the server will update during the transaction handling
- you want to minimize the amount of exchanges

If you use a transaction, you can add your Patient update or patch interaction to it as well.

**Reply by johnkwaters** [reply]
Thanks! Are there any examples of code showing this kind of more complex, real world logic? I fee like the only way I can find out how to do things with the SDK is to ask - more detailed examples would be awesome!

**Reply by mbaltus** [reply]
At the moment, unfortunately we do not have extensive documentation outside the library itself. But for code examples you could take a look at the unit tests. We are also developing a training course to go over several topics, and will probably schedule this in September or October.

**Reply by johnkwaters** [reply]
I have downloaded the source code and am mining the unit tests for examples! Very grateful for all your responses here, very helpful.

**Reply by johnkwaters** [reply]
So this is the code I came up with

The way I reference the Encounter, Patient etc - is this the right way? Feels clunky - I have the objects - is there a helper that will get me the ResourceReferences for them?
And can the TransactionBuilder do this stuff? I.e the get. adding the encounter, adding the observation?


```
var pu = await fc.ReadAsync<Patient>(
    ResourceIdentity.Build("Patient", "ORG-SM-05306f3b-c7c3-4f9a-a441-87feb2ec6a66"));
pu.Address.First().PostalCode = "01008";

var eid = FHIRHelper.CreateId();
var encounter = new Encounter
{
    Id = eid.Value,
    Period = new Period(new FhirDateTime(2022, 5, 15), new FhirDateTime(2022, 5, 15)),
    Status = Encounter.EncounterStatus.Finished,
};
encounter.Identifier.Add(eid);
encounter.Subject = new ResourceReference( $"Patient\\{pu.Id}" );

var oid = FHIRHelper.CreateId();
var observation = new Observation
{
    Id = oid.Value,
    Status = ObservationStatus.Final,
    Code = new CodeableConcept(CodeSystemConstants.LOINC, LOINC.BODY_WEIGHT),
    Value = new Quantity((int)(310/2.2), "kg"),
    Encounter = new ResourceReference($"Encounter\\{encounter.Id}"),
    Subject = new ResourceReference($"Patient\\{pu.Id}")
};
observation.Identifier.Add(oid);

var updated = await fc.UpdateAsync(pu);
```
